File: java/05-reactor/src/main/java/com/fResult/reactor/ReactorApplication.java
package com.fResult.reactor;

import java.util.List;
import java.util.concurrent.ScheduledExecutorService;
import org.springframework.aop.SpringProxy;
import org.springframework.aop.framework.Advised;
import org.springframework.aot.hint.MemberCategory;
import org.springframework.aot.hint.RuntimeHints;
import org.springframework.aot.hint.RuntimeHintsRegistrar;
import org.springframework.aot.hint.TypeReference;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.ImportRuntimeHints;
import org.springframework.core.DecoratingProxy;

@SpringBootApplication
@ImportRuntimeHints(ReactorApplication.Hints.class)
public class ReactorApplication {
  public static void main(String... args) {
    SpringApplication.run(ReactorApplication.class, args);
  }

  static class Hints implements RuntimeHintsRegistrar {
    @Override
    public void registerHints(RuntimeHints hints, ClassLoader classLoader) {
      final var prefix = "reactor.core.publisher.Traces";
      final var classes =
          List.of(
              prefix,
              prefix + "$StackWalkerCallSiteSupplierFactory",
              prefix + "$SharedSecretsCallSiteSupplierFactory",
              prefix + "$ExceptionCallSiteSupplierFactory");

      classes.forEach(
          c -> hints.reflection().registerType(TypeReference.of(c), MemberCategory.values()));

      hints
          .proxies()
          .registerJdkProxy(
              ScheduledExecutorService.class,
              SpringProxy.class,
              Advised.class,
              DecoratingProxy.class);
    }
  }
}

File: kotlin/05-reactor/src/main/kotlin/com/fResult/reactor/ReactorApplication.kt
package com.fResult.reactor

import org.springframework.aop.SpringProxy
import org.springframework.aop.framework.Advised
import org.springframework.aot.hint.MemberCategory
import org.springframework.aot.hint.RuntimeHints
import org.springframework.aot.hint.RuntimeHintsRegistrar
import org.springframework.aot.hint.TypeReference
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.ImportRuntimeHints
import org.springframework.core.DecoratingProxy
import org.springframework.scheduling.concurrent.ScheduledExecutorTask
import reactor.core.publisher.Flux

@SpringBootApplication
@ImportRuntimeHints(Hints::class)
class ReactorApplication

fun main(args: Array<String>) {
  runApplication<ReactorApplication>(*args)
}

class Hints : RuntimeHintsRegistrar {
  override fun registerHints(hints: RuntimeHints, classLoader: ClassLoader?) {
    Flux.just(1)
    val prefix = "reactor.core.publisher.Traces"
    val classes = listOf(
      prefix,
      "$prefix\$StackWalkerCallSiteSupplierFactory",
      "$prefix\$SharedSecretsCallSiteSupplierFactory",
      "$prefix\$ExceptionCallSiteSupplierFactory"
    )

    classes.forEach { hints.reflection().registerType(TypeReference.of(it), *MemberCategory.entries.toTypedArray()) }

    hints.proxies()
      .registerJdkProxy(
        ScheduledExecutorTask::class.java,
        SpringProxy::class.java,
        Advised::class.java,
        DecoratingProxy::class.java
      )
  }
}
